In [1]:
import numpy as np
import pandas as pd

In [19]:
data_dir = "/home/brianmcgloughlin/Japanese_future_real_estate_prices/Data/trade_prices"
# Look at 01 - Hokkaido#

df_01 = pd.read_csv(f'{data_dir}/01.csv')

df_01.info()
df_01.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186238 entries, 0 to 186237
Data columns (total 38 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   No                           186238 non-null  int64  
 1   Type                         186238 non-null  object 
 2   Region                       133069 non-null  object 
 3   MunicipalityCode             186238 non-null  int64  
 4   Prefecture                   186238 non-null  object 
 5   Municipality                 186238 non-null  object 
 6   DistrictName                 183567 non-null  object 
 7   NearestStation               145929 non-null  object 
 8   TimeToNearestStation         144490 non-null  object 
 9   MinTimeToNearestStation      144490 non-null  float64
 10  MaxTimeToNearestStation      140917 non-null  float64
 11  TradePrice                   186238 non-null  int64  
 12  FloorPlan                    20486 non-null   object 
 13 

/tmp/ipykernel_102100/2007917348.py:4: DtypeWarning: Columns (12,36) have mixed types. Specify dtype option on import or set low_memory=False.
  df_01 = pd.read_csv(f'{data_dir}/01.csv')


No                          Type            Region  MunicipalityCode  \
0   1   Residential Land(Land Only)  Residential Area              1101   
1   2   Residential Land(Land Only)  Residential Area              1101   
2   3  Pre-owned Condominiums, etc.               NaN              1101   
3   4   Residential Land(Land Only)  Residential Area              1101   
4   5   Residential Land(Land Only)  Residential Area              1101   

  Prefecture            Municipality DistrictName NearestStation  \
0   Hokkaido  Chuo Ward,Sapporo City   Asahigaoka   Maruyamakoen   
1   Hokkaido  Chuo Ward,Sapporo City   Asahigaoka   Maruyamakoen   
2   Hokkaido  Chuo Ward,Sapporo City   Asahigaoka   Maruyamakoen   
3   Hokkaido  Chuo Ward,Sapporo City   Asahigaoka   Maruyamakoen   
4   Hokkaido  Chuo Ward,Sapporo City   Asahigaoka   Maruyamakoen   

  TimeToNearestStation  MinTimeToNearestStation  MaxTimeToNearestStation  ...  \
0                   28                     28.0                     28.0  ...   
1                   27                     27.0                     27.0  ...   
2                   20                     20.0                     20.0  ...   
3                   26                     26.0                     26.0  ...   
4                   23                     23.0                     23.0  ...   

   Direction Classification  Breadth  \
0      South      City Road     10.2   
1       East      City Road      7.2   
2        NaN            NaN      NaN   
3       West   Private Road      4.0   
4       West      City Road      7.2   

                                        CityPlanning  CoverageRatio  \
0  Category I Exclusively Low-story Residential Zone           40.0   
1  Category I Exclusively Low-story Residential Zone           40.0   
2                       Neighborhood Commercial Zone           80.0   
3  Category I Exclusively Low-story Residential Zone           40.0   
4  Category I Exclusively Low-story Residential Zone           40.0   

   FloorAreaRatio            Period  Year  Quarter  Renovation  Remarks  
0            80.0  4th quarter 2018  2018        4         NaN      NaN  
1            80.0  4th quarter 2018  2018        4         NaN      NaN  
2           200.0  4th quarter 2018  2018        4     Not yet      NaN  
3            80.0  2nd quarter 2018  2018        2         NaN      NaN  
4            80.0  2nd quarter 2018  2018        2         NaN      NaN  

[5 rows x 38 columns]

# My Initial Thoughts/Comments
## Useful

 1.   **Type**                         186238 non-null  object  : This is a string explaining the type of real estate. Useful but maybe hard to work with. Can filter land v building with this.
 
 2.   **Region**                       133069 non-null  object  : Type of area i.e. Commercial / residential / Industrial. Useful but maybe hard to work with.

 4.   **Prefecture**                   186238 non-null  object  : Useful. Probably need to encode. 

 5.   **Municipality**                 186238 non-null  object  : Useful. Can I encode with the municipality code?
 
 6.   **DistrictName**                 183567 non-null  object  : Maybe useful. Difficult to work with

 9.   **MinTimeToNearestStation**      144490 non-null  float64 : Useful.

 10.  **MaxTimeToNearestStation**      140917 non-null  float64 : Useful. Consider taking mean with previous value?

 11.  **TradePrice**                   186238 non-null  int64   : Useful.

 12.  **FloorPlan**                    20486 non-null   object  : Useful. Looks difficult to work with.

 13.  **Area**                         186238 non-null  int64   : Useful.

 14.  **AreaIsGreaterFlag**            186238 non-null  int64   : Useful. Flag if area is larger than 2000 square meters. Convert to bool?

 18.  **Frontage**                     125257 non-null  float64 : Useful.

 19.  **FrontageIsGreaterFlag**        186238 non-null  bool    : Useful. Flag if length of frontage greater than 50m

 20.  **TotalFloorArea**               62490 non-null   float64 : Useful.

 21.  **TotalFloorAreaIsGreaterFlag**  186238 non-null  int64   : Useful. Floor area greater than 2000 sqyare meters. Convert to bool?

 22.  **BuildingYear**                 81078 non-null   float64 : Useful.

 23.  **PrewarBuilding**               186238 non-null  int64   : Useful. True if building year < 1945. Convert to bool?

 25.  **Use**                          81249 non-null   object  : Not so useful for price prediction but useful for filtering data for (non)residential buildings.

 26.  **Purpose**                      56882 non-null   object  : Not so useful for price prediction but useful for filtering data for (non)residential buildings.

 33.  **Period**                       186238 non-null  object  : Useful for timing trends.

 34.  **Year**                         186238 non-null  int64   : Useful for timing trends.

 35.  **Quarter**                      186238 non-null  int64   : Useful for timing trends.


 ## Probably not useful

 **No** - Just an Identifier

 3.   **MunicipalityCode**             186238 non-null  int64   : City code of Japan. Might be useful for encoding.

 7.   **NearestStation**               145929 non-null  object  : Probably matters more if there is a station or not, not which station it is.

 8.   **TimeToNearestStation**         144490 non-null  object  : Messy data. Use min / max values instead.

 15.  **UnitPrice**                    66260 non-null   float64 : Land price per square meter. Redundant since we care about the total trade price. If this is land only or not info is available in Type col.  

 16.  **PricePerTsubo**                66260 non-null   float64 : Probably made redundant given area and tradeprice. 

 17.  **LandShape**                    132674 non-null  object  : Assumed to be unimportant to first order.

 24.  **Structure**                    81430 non-null   object  : Assumed to be unimportant to first order.

 27.  **Direction**                    132631 non-null  object  : Assumed to be unimportant to first order.

 28.  **Classification**               129636 non-null  object  : Claddification of outside road. Probably redundant from region information.

 29.  **Breadth**                      128012 non-null  float64 : Breadth of outside road. Probably redundant from region information.

 30.  **CityPlanning**                 152003 non-null  object  : Assumed to be unimportant to first order.

 31.  **CoverageRatio**                140562 non-null  float64 : Assumed to be unimportant to first order.

 32.  **FloorAreaRatio**               140562 non-null  float64 : Assumed to be unimportant to first order.

 36.  **Renovation**                   19731 non-null   object  : Maybe important but data is sparse.
 
 37.  **Remarks**                      6550 non-null    object  : Sparse. Likely irrelavant in the majority of cases.


In [20]:
# Remove unwanted columns #
df_01.drop(columns=['No', 'FloorPlan', 'Remarks', 'Renovation', 
                     'LandShape', 'Structure', 'Direction', 'Classification', 'Breadth', 
                     'CityPlanning', 'CoverageRatio', 'FloorAreaRatio', 'NearestStation', 'TimeToNearestStation', 'Use', 'PricePerTsubo', 'UnitPrice'], inplace=True)

# Combine MinTimeToNearestStation and MaxTimeToNearestStation into AvgTimeToNearestStation
df_01['AvgTimeToNearestStation'] = df_01[['MinTimeToNearestStation', 'MaxTimeToNearestStation']].mean(axis=1)


# Convert relevant columns to bool type
bool_columns = ['AreaIsGreaterFlag', 'FrontageIsGreaterFlag', 'TotalFloorAreaIsGreaterFlag', 'PrewarBuilding']

for col in bool_columns:
    df_01[col] = df_01[col].astype(bool)



In [21]:
# We only want to investigate real estate that is intended to be used as a house

#Residential_condition = (df_01["Type"] != "Forest Land") & (df_01["Type"] != "Agricultural Land")
Intended_House_condition = (df_01["Purpose"] == "House")

House_df = df_01[Intended_House_condition]

print(f'{len(House_df)} entries, ({(len(House_df) / len(df_01)) * 100} % of total) satisfy the housing condition.')


42410 entries, (22.771936983859362 % of total) satisfy the housing condition.


In [25]:
# Any type that is Land Only will not have a floor size, so we can set the TotalFloorArea to -1. Same logic for Building year and frontage
House_df.loc[House_df['TotalFloorArea'].isna() & (House_df['Type'] == 'Residential Land(Land Only)'), 'TotalFloorArea'] = -1
House_df.loc[House_df['BuildingYear'].isna() & (House_df['Type'] == 'Residential Land(Land Only)'), 'BuildingYear'] = -1
House_df.loc[House_df['Frontage'].isna() & (House_df['Type'] == 'Residential Land(Land Only)'), 'Frontage'] = -1
#If the region or time to station cannot be determined, delete
# Drop rows where 'Region', 'MinTimeToNearestStation', or 'MaxTimeToNearestStation' have NaN values
House_df.dropna(subset=['Region', 'MinTimeToNearestStation', 'MaxTimeToNearestStation'], inplace=True)



/tmp/ipykernel_102100/3291465427.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  House_df.dropna(subset=['Region', 'MinTimeToNearestStation', 'MaxTimeToNearestStation'], inplace=True)


In [26]:
# Count rows with NaN values in the entire DataFrame
nan_rows_count = House_df.isna().sum(axis=1)
rows_with_nan = nan_rows_count[nan_rows_count > 0]

# Output the number of rows with NaN values
print(f'Number of rows with NaN values: {len(rows_with_nan)}')

Number of rows with NaN values: 1984


In [33]:
# Set pandas display options to show all rows and columns
pd.set_option('display.max_rows', 200)  # Remove row limit
pd.set_option('display.max_columns', 25)  # Remove column limit
House_df

Type            Region  \
1               Residential Land(Land Only)  Residential Area   
4               Residential Land(Land Only)  Residential Area   
8               Residential Land(Land Only)  Residential Area   
12      Residential Land(Land and Building)  Residential Area   
20              Residential Land(Land Only)  Residential Area   
...                                     ...               ...   
179713  Residential Land(Land and Building)   Commercial Area   
179730  Residential Land(Land and Building)  Residential Area   
179734  Residential Land(Land and Building)  Residential Area   
179736  Residential Land(Land and Building)   Commercial Area   
179777          Residential Land(Land Only)  Residential Area   

        MunicipalityCode Prefecture                     Municipality  \
1                   1101   Hokkaido           Chuo Ward,Sapporo City   
4                   1101   Hokkaido           Chuo Ward,Sapporo City   
8                   1101   Hokkaido           Chuo Ward,Sapporo City   
12                  1101   Hokkaido           Chuo Ward,Sapporo City   
20                  1101   Hokkaido           Chuo Ward,Sapporo City   
...                  ...        ...                              ...   
179713              1668   Hokkaido  Shiranuka Town,Shiranuka County   
179730              1668   Hokkaido  Shiranuka Town,Shiranuka County   
179734              1668   Hokkaido  Shiranuka Town,Shiranuka County   
179736              1668   Hokkaido  Shiranuka Town,Shiranuka County   
179777              1668   Hokkaido  Shiranuka Town,Shiranuka County   

            DistrictName  MinTimeToNearestStation  MaxTimeToNearestStation  \
1             Asahigaoka                     27.0                     27.0   
4             Asahigaoka                     23.0                     23.0   
8             Asahigaoka                     29.0                     29.0   
12            Asahigaoka                     28.0                     28.0   
20            Asahigaoka                     24.0                     24.0   
...                  ...                      ...                      ...   
179713  Higashi1jominami                      4.0                      4.0   
179730  Higashi2jominami                      4.0                      4.0   
179734  Higashi2jominami                      2.0                      2.0   
179736  Higashi2jominami                      2.0                      2.0   
179777      (No Address)                     30.0                     60.0   

        TradePrice  Area  AreaIsGreaterFlag  Frontage  FrontageIsGreaterFlag  \
1         38000000   310              False      21.5                  False   
4         25000000   430              False      16.0                  False   
8         20000000   165              False      12.6                  False   
12        40000000   400              False      14.0                  False   
20        23000000   370              False      -1.0                  False   
...            ...   ...                ...       ...                    ...   
179713     1600000   250              False       9.1                  False   
179730     6600000   190              False      10.0                  False   
179734      570000    95              False       6.6                  False   
179736     2000000   210              False       8.9                  False   
179777      950000  2000               True      34.0                  False   

        TotalFloorArea  TotalFloorAreaIsGreaterFlag  BuildingYear  \
1                 -1.0                        False          -1.0   
4                 -1.0                        False          -1.0   
8                 -1.0                        False          -1.0   
12               190.0                        False        1989.0   
20                -1.0                        False          -1.0   
...                ...                          ...           ...   
179713      

In [34]:
#Drop remaining NaN values
House_df.dropna(inplace=True)


/tmp/ipykernel_102100/2332581553.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  House_df.dropna(inplace=True)


In [35]:
House_df

Type            Region  \
1               Residential Land(Land Only)  Residential Area   
4               Residential Land(Land Only)  Residential Area   
8               Residential Land(Land Only)  Residential Area   
12      Residential Land(Land and Building)  Residential Area   
20              Residential Land(Land Only)  Residential Area   
...                                     ...               ...   
179713  Residential Land(Land and Building)   Commercial Area   
179730  Residential Land(Land and Building)  Residential Area   
179734  Residential Land(Land and Building)  Residential Area   
179736  Residential Land(Land and Building)   Commercial Area   
179777          Residential Land(Land Only)  Residential Area   

        MunicipalityCode Prefecture                     Municipality  \
1                   1101   Hokkaido           Chuo Ward,Sapporo City   
4                   1101   Hokkaido           Chuo Ward,Sapporo City   
8                   1101   Hokkaido           Chuo Ward,Sapporo City   
12                  1101   Hokkaido           Chuo Ward,Sapporo City   
20                  1101   Hokkaido           Chuo Ward,Sapporo City   
...                  ...        ...                              ...   
179713              1668   Hokkaido  Shiranuka Town,Shiranuka County   
179730              1668   Hokkaido  Shiranuka Town,Shiranuka County   
179734              1668   Hokkaido  Shiranuka Town,Shiranuka County   
179736              1668   Hokkaido  Shiranuka Town,Shiranuka County   
179777              1668   Hokkaido  Shiranuka Town,Shiranuka County   

            DistrictName  MinTimeToNearestStation  MaxTimeToNearestStation  \
1             Asahigaoka                     27.0                     27.0   
4             Asahigaoka                     23.0                     23.0   
8             Asahigaoka                     29.0                     29.0   
12            Asahigaoka                     28.0                     28.0   
20            Asahigaoka                     24.0                     24.0   
...                  ...                      ...                      ...   
179713  Higashi1jominami                      4.0                      4.0   
179730  Higashi2jominami                      4.0                      4.0   
179734  Higashi2jominami                      2.0                      2.0   
179736  Higashi2jominami                      2.0                      2.0   
179777      (No Address)                     30.0                     60.0   

        TradePrice  Area  AreaIsGreaterFlag  Frontage  FrontageIsGreaterFlag  \
1         38000000   310              False      21.5                  False   
4         25000000   430              False      16.0                  False   
8         20000000   165              False      12.6                  False   
12        40000000   400              False      14.0                  False   
20        23000000   370              False      -1.0                  False   
...            ...   ...                ...       ...                    ...   
179713     1600000   250              False       9.1                  False   
179730     6600000   190              False      10.0                  False   
179734      570000    95              False       6.6                  False   
179736     2000000   210              False       8.9                  False   
179777      950000  2000               True      34.0                  False   

        TotalFloorArea  TotalFloorAreaIsGreaterFlag  BuildingYear  \
1                 -1.0                        False          -1.0   
4                 -1.0                        False          -1.0   
8                 -1.0                        False          -1.0   
12               190.0                        False        1989.0   
20                -1.0                        False          -1.0   
...                ...                          ...           ...   
179713      